gh-75229: make ensurepip honour value of --prefix option

mcepl opened the pull request:
When cross-compiling, the local Python interpreter that is used to run `ensurepip` may not have the same value of `sys.prefix` as the value of the 'prefix' variable that is set in the Makefile.

With the following values used to install Python locally for a later copy to the files hierarchy owned by the 'termux' application on an Android device:

```
    DESTDIR=/tmp/android
    prefix=/data/data/com.termux/files/usr/local
```

'make install' causes ensurepip to install pip in `$(DESTDIR)/usr/local` instead of the expected `$(DESTDIR)/$(prefix)` where is installed the standard library.

The attached patch fixes the problem. The patch was implemented assuming that pip uses `distutils` for the installation (note that setup.py also uses the --prefix option in the Makefile), but I know nothing about pip so forgive me if the patch is wrong and please just assume it is just a way to demonstrate the problem.

Fixes: https://github.com/python/cpython/issues/75229
Co-authored-by: Pradyun Gedam <pradyunsg@gmail.com>
Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>
Co-authored-by: Zackery Spytz <zspytz@gmail.com>
References: https://github.com/python/cpython/pull/17634
Signed-off-by: Matěj Cepl <mcepl@cepl.eu>

<!-- gh-issue-number: gh-75229 -->
* Issue: gh-75229
<!-- /gh-issue-number -->

xdegaye committed:
bpo-31046: ensurepip does not honour the value of $(prefix)

When cross-compiling, the local Python interpreter that is used
to run `ensurepip` may not have the same value of `sys.prefix` as
the value of the 'prefix' variable that is set in the Makefile.

With the following values used to install Python locally for
a later copy to the files hierarchy owned by the 'termux'
application on an Android device:

    DESTDIR=/tmp/android
    prefix=/data/data/com.termux/files/usr/local

'make install' causes ensurepip to install pip in
$(DESTDIR)/usr/local instead of the expected $(DESTDIR)/$(prefix)
where is installed the standard library.

The attached patch fixes the problem. The patch was implemented
assuming that pip uses distutils for the installation (note that
setup.py also uses the --prefix option in the Makefile), but I
know nothing about pip so forgive me if the patch is wrong and
please just assume it is just a way to demonstrate the problem.

Fixes: https://github.com/python/cpython/issues/75229
Fixes: https://bugs.python.org/issue31046
Co-authored-by: Pradyun Gedam <pradyunsg@gmail.com>
Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>
Co-authored-by: Zackery Spytz <zspytz@gmail.com>
References: https://github.com/python/cpython/pull/17634
Signed-off-by: Matěj Cepl <mcepl@cepl.eu>

mcepl committed:
bpo-31046: Fix ensurepip script shebangs with --root

When using `python -m ensurepip` with the `--root` option for staged
installations, the generated pip script contained an incorrect shebang
that pointed into the staging directory. This made the installation
unusable once the staging directory was removed.

This commit fixes the issue by using the internal pip `--executable`
option to force the shebang to point to the correct, final
interpreter path.

It also corrects related pathing issues:

- Removes the check that incorrectly disallowed using --root and
  --prefix together.
- Defaults the installation prefix to `/` when --root is used alone,
  ensuring installation occurs at the base of the staging directory.

References: https://github.com/python/cpython/pull/17634#discussion_r1622453325
Signed-off-by: Matěj Cepl <mcepl@cepl.eu>

JasonGantner commented on Lib/ensurepip/__init__.py:
```diff
@@ -160,15 +160,31 @@ def _bootstrap(*, root=None, upgrade=False, user=False,
+            # /usr/local, and install into ${root}/usr/local/.
+            effective_prefix = prefix
+            if root and not prefix:
+                effective_prefix = "/"
```
Usual use cases for `ensurepip` are:
- installing pip on the current(host) system -> `sys.prefix` __is__ the right prefix to use
- installing pip in a packaging directory -> the Makefile _always_ provides the prefix (given `Makefile.pre.in` is patched correctly by this PR)

Adding this override of the default `prefix` would break the first use case without clear benefit to the second.

JasonGantner replied:
This issue remains

JasonGantner commented on Makefile.pre.in:
```diff
@@ -2336,7 +2336,7 @@ install: @FRAMEWORKINSTALLFIRST@ @INSTALLTARGETS@ @FRAMEWORKINSTALLLAST@
 		esac; \
 		$(RUNSHARED) $(PYTHON_FOR_BUILD) -m ensurepip \
-			$$ensurepip --root=$(DESTDIR)/ ; \
+			$$ensurepip --prefix=$(prefix) ; \
```
We must preserve the `--root` argument here.

```suggestion
			$$ensurepip --root=$(DESTDIR)/ --prefix=$(prefix) ; \
```

mcepl replied:
OK

JasonGantner commented on Makefile.pre.in:
```diff
@@ -2347,7 +2347,7 @@ altinstall: commoninstall
 		esac; \
 		$(RUNSHARED) $(PYTHON_FOR_BUILD) -m ensurepip \
-			$$ensurepip --root=$(DESTDIR)/ ; \
+			$$ensurepip --prefix=$(prefix) ; \
```
and also here

```suggestion
			$$ensurepip --root=$(DESTDIR)/ --prefix=$(prefix) ; \
```

mcepl replied:
OK

JasonGantner reviewed (changes requested):
Thank you for the work improving this PR.

I've noticed two small issues, could you take them into consideration?

PS: the issue in `Makefile.pre.in` is the reason a lot of tests are failing

mcepl committed:
don't throw away `--root` parameter of ensurepip

Don't break the default installation on the current system with good `sys.prefix`.

JasonGantner commented on Lib/test/test_ensurepip.py:
```diff
@@ -100,6 +100,21 @@ def test_bootstrapping_with_root(self):
+        with self.assertRaises(ValueError):
+            ensurepip.bootstrap(root="", prefix="")
+        self.assertFalse(self.run_pip.called)
+
```
This test must be removed since `root` and `prefix` aren't mutually exclusive.

mcepl replied:
D'oh!

mcepl committed:
fix: remove test_root_and_prefix_mutual_exclusive

It is made incorrect by the previous commit.

mcepl committed:
Remove &nbsp; characters.

mcepl committed:
Fix ensurepip tests to match new --executable argument behavior

Update test expectations in test_ensurepip.py to account for the new
--executable argument that gets added when using --prefix or --root
options. The implementation now correctly sets the executable path
when installing with a custom prefix, but the tests weren't updated
to expect this additional argument.

- test_bootstrapping_with_root: expect --prefix "/" and --executable
- test_bootstrapping_with_prefix: expect --executable argument

mcepl committed:
gh-75229: limit ensurepip prefix overrides to explicit requests

Keep root-only installs using the interpreter's default prefix while
preserving the cross-build fix for Makefile-driven installs that pass an
explicit prefix.

mcepl committed:
fix ensurepip shebangs via in-process pip monkey-patching

Replace the unsupported --executable pip argument with an in-process
monkey-patch of pip's internal script generation logic.

The bundled version of pip does not support the --executable flag,
causing integration failures (e.g., during 'make install'). This change
intercepts the --executable argument in _run_pip and applies it by
patching pip._internal.operations.install.wheel and
pip._vendor.distlib.scripts.ScriptMaker before the installation starts.

This ensures that scripts installed via ensurepip (with --prefix or
--root) use the correct target interpreter shebang even when
cross-compiling or performing staged installs.
